core/src/main/java/com/linecorp/armeria/server/annotation/JavaBeanRequestConverterFunction.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.annotation;
+
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.convertParameter;
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.httpParametersOf;
+import static org.reflections.ReflectionUtils.getAllFields;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpParameters;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.util.Holder;
+import com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.EnumConverter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.AsciiString;
+
+public class JavaBeanRequestConverterFunction implements RequestConverterFunction {
+    private static final Map<Class<?>, BeanParamInfo> beanParamsMap = new HashMap<>();
+
+    private static final List<Class<?>> ignoreResultTypeList = new ArrayList<>();
+
+    @Override
+    public Object convertRequest(final ServiceRequestContext ctx,
+                                 final AggregatedHttpMessage request,
+                                 final Class<?> expectedResultType) throws Exception {
+        final BeanParamInfo beanParamInfo = beanParamsMap.get(expectedResultType);
+
+        if (beanParamInfo == null) {
+            // no BeanParamInfo for this result type, we should not handled it
+            return RequestConverterFunction.fallthrough();
+        }
+
+        if (!shouldHandle(request)) {
```

Comment:
I'm not sure what if we remove `shouldHandle()`? We collected `expectedResultType`s which we can handle, so I think we should handle the request if `beanParamInfo` is non-null.

Reply:
The current implementation is, `shouldHandle` also checks if contentType of request is in JSON format, if it is, this request should be handled by `JacksonRequestConverterFunction` .

In this strategy, we can re-use same Java Bean for JSON request(`JacksonRequestConverterFunction`) and normal request`JavaBeanRequestConverterFunction`


If we remove this `shouldHandle` check, then there will be not this `re-use same Java Bean` feature.

So, what do you think, which is better?

Reply:
I can see what you intend. Is that one of your usage? 

I'm worried that your strategy might make a user be confused. Let's assume that a user writes `@Param` and `@Header` annotation in their class for extracting HTTP request information. In this case, it would not be converted if the request has 'json' as its body. There's no way to do it except that they change their request body.

Actually, the other request converts are originally designed to handle the body of the request, but I think this request converter is different to the others. If I understand correctly, its goal is to provide convenience to a user to get a parameter or a header from the request easily. 

So, I think that the case you mentioned might be an advanced use case. How about:
- making `shouldHandle` `protected` and the default implementation always return `true`
  (or, adding a `Function` type member as a decision maker?)
- removing `JavaBeanRequestConverterFunction` from the default converters
- adding it by default only if a user doesn't specify it as their request converter

It makes our implementation more complicated, but it makes a user be able to customize the converter. 

Please let me know what you think. Also, let's wait what our folks think. :-)

Reply:
hmm....
That sound like too complicated, and hard to understand.
Let's just remove the `shouldHandle` check, and keep it simple for now.
If this kind of feature request exits in the future, let's deal with it then

Reply:
fixed

Reply:
I thought you need to reuse your Java Bean. :-)
Thank you for the change.